### PR TITLE
initial version of docs for testing

### PR DESCRIPTION
**Pull request by mariodavid**

### CODE REVIEW

File: content/modules/testing/pages/index.adoc [hunk 1]
```diff
@@ -0,0 +1,9 @@
+= Testing
+
+Automated Testing is a core part of Jmix as it allows creating long-term maintainable applications. Jmix provides various capabilities to support the application developer to create different types of automated tests.
+
+Since Jmix is based on Spring Boot, all Spring Boot Testing capabilities also apply to Jmix.
+
+Spring has comprehensive documentation in their https://docs.spring.io/spring-framework/docs/current/reference/html/testing.html[Spring Framework: Testing reference documentation] as well as in Spring Boot's https://docs.spring.io/spring-boot/docs/current/reference/html/features.html#features.testing[Testing feature documentation].
```
Comment (knstvk) [inline content/modules/testing/pages/index.adoc]: The link https://docs.spring.io/spring-framework/docs/current/reference/html/testing.html doesn't go where intended.

Also, add ^ to all external links to open in a new tab, e.g. https://address[Title^].

File: content/modules/testing/pages/getting-started.adoc [hunk 2]
```diff
@@ -0,0 +1,82 @@
+= Getting Started with Testing
+
+This section contains a practical guide on getting started with creating and running the first test case with Jmix.
+
+By default, every Jmix project already contains the directory structure for automated tests. It also scaffolds one automated test case `UserTest` in the directory: `src/test/java/\<<project-base-package>>/user/UserTest.java` that acts as an example for to create your own test classes.
```
Comment (knstvk) [inline content/modules/testing/pages/getting-started.adoc]: Let's assume this guide talks about v.1.5 and classic UI. So there are 2 tests in a new project created with "Full-stack application (Java)" template: UserTest and UserUiTest.

File: content/modules/testing/pages/unit-tests.adoc [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+
+Mockito is a mocking library that supports this kind of emulation. Jmix includes Mockito by default.
+
+Here is an example on how this test case could look like:
+
+[source,java,indent=0]
+----
+include::example$/ex1/src/test/java/testing/ex1/order/RecentOrdersCounterTest.java[tags=recent-order-counter-test]
+----
+<1> the `mock` method creates an emulation instance that we can use to control the behaviour of the class.
```
Comment (knstvk) [inline content/modules/testing/pages/unit-tests.adoc]: Start with capital letter

**Review by knstvk — COMMENTED** [review]
Great start, looking forward to the continuation!

File: content/modules/testing/pages/index.adoc (hunk 1 above)
Comment (mariodavid) [reply content/modules/testing/pages/index.adoc]: done

File: content/modules/testing/pages/integration-tests.adoc [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+public class CustomerServiceTest {
+
+    @Autowired
+    CustomerService customerService;
+
+    // ...
+}
+----
+
+== Database interactions
```
Comment (knstvk) [inline content/modules/testing/pages/integration-tests.adoc]: Use Title Case. Check all titles please.

File: content/modules/testing/pages/integration-tests.adoc [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+
+In an integration test there are two main reasons to interact with the database. The first reason is that we want to set up test data. This is oftentimes required as a prerequisite for the execution of the test case. For this we can use the regular Jmix functionalities like `DataManager` to interact with the database as we would in production code.
+
+The second reason is that indirect: by executing the business logic that we want to test, this business logic perform some operation against the database. Let's look at an example for both of those scenarios:
```
Comment (knstvk) [inline content/modules/testing/pages/integration-tests.adoc]: "The second reason is that indirect" sounds strange.

File: content/modules/testing/pages/integration-tests.adoc [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+
+[source,java,indent=0,title='CustomerServiceTest.java']
+----
+include::example$/ex1/src/test/java/testing/ex1/customer/CustomerServiceTest.java[tags=find-by-email-found]
+----
+<1> The `DataManager` is used in the test, to store a test customer in the database.
+<2> The `CustomerService` is used to perform a database lookup of customers by Email.
```
Comment (knstvk) [inline content/modules/testing/pages/integration-tests.adoc]: "email". No need to capitalize.

File: content/modules/testing/pages/integration-tests.adoc [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+
+There are several ways on how to clean up the test data. The first is to keep references of data that was created during the test. In the example above, we could keep a reference to the customer that was created in the test and delete it after the test has finished via `dataManager.remove(customer)`. This is a valid approach, but it requires some additional code in the test. Additionally, it is not always possible to keep a reference to the data that was created during the test. For example, if we create a new entity in the production code, we don't have a reference to it in the test. Furthermore, in case of an exception during the test, the cleanup code might not be executed.
```
Comment (knstvk) [inline content/modules/testing/pages/integration-tests.adoc]: "data that were created"

**Review by knstvk — COMMENTED** [review]
A few minor corrections.

**Comment by knstvk** [comment]
Due to the latest changes in the docs branches, you need to make the PR for the release_1_5 branch.

File: content/modules/testing/pages/getting-started.adoc (hunk 2 above)
Comment (mariodavid) [reply content/modules/testing/pages/getting-started.adoc]: done

File: content/modules/testing/pages/unit-tests.adoc (hunk 3 above)
Comment (mariodavid) [reply content/modules/testing/pages/unit-tests.adoc]: done

File: content/modules/testing/pages/integration-tests.adoc (hunk 4 above)
Comment (mariodavid) [reply content/modules/testing/pages/integration-tests.adoc]: done

File: content/modules/testing/pages/integration-tests.adoc (hunk 6 above)
Comment (mariodavid) [reply content/modules/testing/pages/integration-tests.adoc]: done

File: content/modules/testing/pages/integration-tests.adoc (hunk 7 above)
Comment (mariodavid) [reply content/modules/testing/pages/integration-tests.adoc]: can you elaborate what you mean here?

Comment (knstvk) [reply content/modules/testing/pages/integration-tests.adoc]: I just heard that "data" is more often used as plural. Not important though.

**Comment by mariodavid** [comment]
closed in favour of https://github.com/jmix-framework/jmix-docs/pull/560

**Pull request closed without merging**